### ISSUE TITLE
Bloc.add is not called in initState() of a stateful widget

### ISSUE THREAD

**Issue by thomasycf** [felangel/bloc#936]
Hi there,

I use a bloc to control displaying of search history and search results in a page. I make it to display the search history when it is open. It works properly as expected when I open the page at first time. However, if I leave the page from search results page and later re-open the search page, rather than going to the search history page as expected, it go to the search results page with what the state I got last time.
The log showing the SearchPage.initState() is called  but _searchBodyBloc.add(DefaultPageEvent()) is not as I tell that from mapEventToState().

2nd screen shot showing '_searchBodyBloc.add(DefaultPageEvent())' is not called.

Any idea what is causing the problem?

Work around is to call _searchBodyBloc.add(DefaultPageEvent()) in SearchBody.dispose() but I think this is abnormal.

SearchPage() - the parent
 initState(){
_searchBodyBloc.add(DefaultPageEvent()); 
}

SearchBody() - the child (displaying search history / search results / CircularProgressIndicator)

SearchBodyBloc mapEventToState:
   if (event is DefaultPageEvent) {
      yield InitialSearchBodyState(); //display a blank Container()
// get search history from sqlite
      List<Searchlog> _searchlogList = await _searchlogDao.getAllSearchlogs();
      yield DefaultSearchState(_searchlogList); // display search history
    }

**Screenshots**
![Screenshot from 2020-03-04 15-37-01](https://user-images.githubusercontent.com/37063910/75856656-c16f6b80-5e2f-11ea-838a-b167629dcebf.png)
![Screenshot from 2020-03-04 15-51-15](https://user-images.githubusercontent.com/37063910/75856771-009dbc80-5e30-11ea-8c19-1a01c5b674e0.png)

**Comment by felangel** [comment]
Hi @thomasycf 👋
Thanks for opening an issue!

Are you able to share a link to a sample app which illustrates the issue? I’m guessing you’re not scoping the bloc to just that screen so when you leave and come back add is called but no state changes happens because you’re yielding the same state as the bloc’s current state.

**Comment by felangel** [comment]
Closing for now but feel free to comment with additional details or a link to a sample app and I'm happy to continue the conversation 👍

**Comment by thomasycf** [comment]
Hi Felix
Thanks for looking into this.  I think I found the cause of the issue. It did not work properly when I used BlocProvider with a singleton Bloc object create from factory(). Is this normal? But I never dispose the Bloc when I leave the page.

I have created a project to simulate the problem as below.  If you run the sample, click to go to 'Search Page', then go to 'Result Page', then just press phone's back button. It would leave the page but if you re-enter the Search Page, you would find the Bloc fails to work.

https://github.com/thomasycf/bloc_blocProvider.git
